src/database.ts

```diff
@@ -3346,6 +3349,91 @@ class Database extends common.GrpcServiceObject {
     return proxyStream as NodeJS.ReadableStream;
   }
 
+  /**
+   * Apply Blind Write of the Mutations
+   *
+   * writeAtLeastOnce(Blind Write) requests are not replay protected, meaning that it may apply mutations more
+   * than once, if the mutations are not idempotent, this may lead to a failure being
+   * reported when the mutation was applied once. Replays of non-idempotent mutations may
+   * have undesirable effects. For example, replays of an insert mutation may produce an
+   * already exists error. For this reason, most users of the library will prefer to use
+   * {@link runTransaction} instead.
+   *
+   * However, {@link writeAtLeastOnce()} requires only a single RPC, whereas {@link runTransaction()}
+   * requires two RPCs (one of which may be performed in advance), and so this method may be
+   * appropriate for latency sensitive and/or high throughput blind writing.
+   *
+   * We recommend structuring your mutation groups to be idempotent to avoid this issue.
+   *
+   * @param {Mutation} [mutation]  Mutations to be applied.
+   * @param {CallOptions} [options] Options object for blind write request.
+   * @param {CommitCallback} [callback] Callback function for blind write request.
+   *
+   * @returns {Promise}
+   *
+   * @example
+   * ```
+   * const {Spanner} = require('@google-cloud/spanner');
+   * const spanner = new Spanner();
+   *
+   * const instance = spanner.instance('my-instance');
+   * const database = instance.database('my-database');
+   * const mutation = new Mutation();
+   * mutation.insert('Singers', {
+   *  SingerId: '1',
+   *  FirstName: 'Marc',
+   *  LastName: 'Richards',
+   *  });
+   * mutation.update('Singers', {
+   *  SingerId: '1',
+   *  FirstName: 'John',
+   *  LastName: 'Richards',
+   *  });
+   *
+   * try {
+   *  const [response, err] = await database.writeAtLeastOnce(mutation, {});
+   *  console.log(response.commitTimestamp);
+   * } catch(err) {
+   *  console.log("Error: ", err);
+   * }
+   * ```
+   */
+  writeAtLeastOnce(mutation: Mutation): Promise<CommitResponse>;
+  writeAtLeastOnce(
+    mutation: Mutation,
+    options: CallOptions
+  ): Promise<CommitResponse>;
+  writeAtLeastOnce(mutation: Mutation, callback: CommitCallback): void;
+  writeAtLeastOnce(
+    mutation: Mutation,
+    optionsOrCallback?: CallOptions | CommitCallback,
+    callback?: CommitCallback
+  ): void | Promise<CommitResponse> {
+    const cb =
+      typeof optionsOrCallback === 'function'
+        ? (optionsOrCallback as CommitCallback)
+        : callback;
+    const options =
+      typeof optionsOrCallback === 'object' && optionsOrCallback
+        ? (optionsOrCallback as CallOptions)
+        : {};
+    this.pool_.getSession((err, session?, transaction?) => {
+      if (err && isSessionNotFoundError(err as grpc.ServiceError)) {
+        cb
+          ? this.writeAtLeastOnce(mutation, cb)
+          : this.writeAtLeastOnce(mutation, options);
```

Comment:
Can't this method also be called with __both__ options and a callback? It seems that in that case we are ignoring the options.

Reply:
Typescript not able to infer the correct types from the overload signature. Hence we need to make it clear to typescript that correct type is pass

Reply:
Yeah, that was not my point. My point was that the original method call could be something like this:

```
database.writeAtLeastOnce(myMutations, myOptions, myCallback);
```

That then leads to the line that we are talking about here calling the following:

```
        cb
          ? this.writeAtLeastOnce(mutations, cb)
          : this.writeAtLeastOnce(mutations, options);
```

And as __both__ options and a callback was passed in as an argument, the ternary will resolve to:

```
this.writeAtLeastOnce(mutations, cb)
```

Put another way:
1. If the original method is called with __both__ options and a callback.
2. And a `SessionNotFound` error occurs.
3. Then the actual method call that is being executed during the retry is __without__ options.

Reply:
Done